src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -564,11 +564,20 @@ public void setUserProperties(final JSONObject userProperties, final boolean rep
         runOnLogThread(new Runnable() {
             @Override
             public void run() {
-                Iterator<?> keys = userProperties.keys();
+                // Create deep copy of input userProperties to prevent ConcurrentModification
+                final JSONObject copy;
+                try {
+                    copy = new JSONObject(userProperties.toString());
```

Comment:
Isn't there still a risk of ConcurrentModification cause you're calling toString here?